src/psyclone/domain/lfric/arg_index_to_metadata_index.py

```diff
@@ -0,0 +1,249 @@
+
+'''This module implements a class that provides a mapping from the
+index of a kernel argument to the index of its associated meta_arg
+kernel metadata.
+
+'''
+from psyclone.domain.lfric.metadata_to_arguments_rules import \
+    MetadataToArgumentsRules
+
+# pylint: disable=too-few-public-methods
+
+
+class ArgIndexToMetadataIndex(MetadataToArgumentsRules):
+    '''Provides a mapping from the index of a kernel argument to the index
+    of its associated meta_arg kernel metadata. Mappings are not
+    provided to arguments that do not have associated meta_arg kernel
+    metadata.
+
+    '''
+
+    @classmethod
+    def _initialise(cls, _):
+        '''Initial state (for the _info variable) is not used by this subclass
+        i.e. the optional info argument to the mapping method is not
+        used. Instead, initial state is always set to an empty
+        dictionary. Also initialises any additional state for this
+        class.
+
+        '''
+        super()._initialise({})
+        cls._index = 0
+
+    @classmethod
+    def _scalar(cls, meta_arg):
+        '''Argument providing an LFRic scalar value.
+
+        :param meta_arg: the metadata associated with this scalar argument.
+        :type meta_arg: \
+            :py:class:`psyclone.domain.lfric.kernel.ScalarArgMetadata`
+
+        '''
+        cls._add_arg(meta_arg)
+
+    @classmethod
+    def _field(cls, meta_arg):
+        '''Argument providing an LFRic field.
+
+        :param meta_arg: the metadata associated with this field argument.
+        :type meta_arg: \
+            :py:class:`psyclone.domain.lfric.kernel.FieldArgMetadata`
+
+        '''
+        cls._add_arg(meta_arg)
+
+    @classmethod
+    def _field_vector(cls, meta_arg):
+        '''Arguments providing the components of an LFRic field vector.
+
+        :param meta_arg: the metadata associated with this field \
+            vector argument.
+        :type meta_arg: \
+            :py:class:`psyclone.domain.lfric.kernel.FieldVectorArgMetadata`
+
+        '''
+        for _ in range(int(meta_arg.vector_length)):
+            cls._add_arg(meta_arg)
+
+    @classmethod
+    def _operator(cls, meta_arg):
+        '''Arguments providing an LMA operator.
+
+        :param meta_arg: the metadata associated with the operator \
+            arguments.
+        :type meta_arg: \
+            :py:class:`psyclone.domain.lfric.kernel.OperatorArgMetadata`
+
+        '''
+        cls._add_arg(meta_arg)
+
+    @classmethod
+    def _cma_operator(cls, meta_arg):
+        '''Arguments providing a columnwise operator.
+
+        :param meta_arg: the metadata associated with the CMA operator \
+            arguments.
+        :type meta_arg: :py:class:`psyclone.domain.lfric.kernel.\
+            ColumnwiseOperatorArgMetadata`
+
+        '''
+        cls._add_arg(meta_arg)
+
+    @classmethod
+    def _add_arg(cls, meta_arg):
+        '''Utility method to provide a map from the current argument index
```

Comment:
Please update: i.e. "adds an entry for the supplied meta_arg..."

Reply:
Re-written, but slightly differently to what was suggested.